File: backend/security.py
"""Password hashing + JWT token helpers."""
from datetime import datetime, timezone
import bcrypt
import jwt
from fastapi import Response

from config import JWT_SECRET, JWT_ALGORITHM, ACCESS_TTL, REFRESH_TTL, COOKIE_SECURE


def hash_password(pw: str) -> str:
    return bcrypt.hashpw(pw.encode("utf-8"), bcrypt.gensalt()).decode("utf-8")


def verify_password(pw: str, hashed: str) -> bool:
    return bcrypt.checkpw(pw.encode("utf-8"), hashed.encode("utf-8"))


def create_access_token(user_id: str, email: str, role: str) -> str:
    payload = {
        "sub": user_id, "email": email, "role": role, "type": "access",
        "exp": datetime.now(timezone.utc) + ACCESS_TTL,
    }
    return jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)


def create_refresh_token(user_id: str) -> str:
    payload = {
        "sub": user_id, "type": "refresh",
        "exp": datetime.now(timezone.utc) + REFRESH_TTL,
    }
    return jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)


def decode_token(token: str) -> dict:
    return jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])


def set_auth_cookies(response: Response, access: str, refresh: str) -> None:
    samesite = "none" if COOKIE_SECURE else "lax"
    response.set_cookie("access_token", access, httponly=True, secure=COOKIE_SECURE,
                        samesite=samesite, max_age=int(ACCESS_TTL.total_seconds()), path="/")
    response.set_cookie("refresh_token", refresh, httponly=True, secure=COOKIE_SECURE,
                        samesite=samesite, max_age=int(REFRESH_TTL.total_seconds()), path="/")


def clear_auth_cookies(response: Response) -> None:
    response.delete_cookie("access_token", path="/")
    response.delete_cookie("refresh_token", path="/")

File: backend-node/src/security.js
// Password hashing + JWT token helpers.
"use strict";
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");
const config = require("./config");

function hashPassword(pw) {
  return bcrypt.hashSync(pw, 10);
}

function verifyPassword(pw, hashed) {
  return bcrypt.compareSync(pw, hashed);
}

function createAccessToken(userId, email, role) {
  return jwt.sign(
    { sub: userId, email, role, type: "access" },
    config.JWT_SECRET,
    { algorithm: config.JWT_ALGORITHM, expiresIn: config.ACCESS_TTL_SECONDS },
  );
}

function createRefreshToken(userId) {
  return jwt.sign(
    { sub: userId, type: "refresh" },
    config.JWT_SECRET,
    { algorithm: config.JWT_ALGORITHM, expiresIn: config.REFRESH_TTL_SECONDS },
  );
}

function decodeToken(token) {
  return jwt.verify(token, config.JWT_SECRET, { algorithms: [config.JWT_ALGORITHM] });
}

// CSRF-protects the Google OAuth redirect round-trip: signed so it can't be
// forged, short-lived so a leaked/logged URL can't be replayed later.
function createOAuthState() {
  return jwt.sign({ purpose: "google_oauth" }, config.JWT_SECRET, {
    algorithm: config.JWT_ALGORITHM, expiresIn: 300,
  });
}
function verifyOAuthState(state) {
  const payload = jwt.verify(state, config.JWT_SECRET, { algorithms: [config.JWT_ALGORITHM] });
  if (payload.purpose !== "google_oauth") throw new Error("Invalid state token");
}

function setAuthCookies(res, access, refresh) {
  const sameSite = config.COOKIE_SECURE ? "none" : "lax";
  res.cookie("access_token", access, {
    httpOnly: true, secure: config.COOKIE_SECURE, sameSite,
    maxAge: config.ACCESS_TTL_SECONDS * 1000, path: "/",
  });
  res.cookie("refresh_token", refresh, {
    httpOnly: true, secure: config.COOKIE_SECURE, sameSite,
    maxAge: config.REFRESH_TTL_SECONDS * 1000, path: "/",
  });
}

function clearAuthCookies(res) {
  res.clearCookie("access_token", { path: "/" });
  res.clearCookie("refresh_token", { path: "/" });
}

module.exports = {
  hashPassword, verifyPassword,
  createAccessToken, createRefreshToken, decodeToken,
  createOAuthState, verifyOAuthState,
  setAuthCookies, clearAuthCookies,
};
